### PR TITLE
Add Ruby Globals cheat sheet

### PR DESCRIPTION
**Pull request by kotoshenya**
**What does your Instant Answer do?**
Displays a list of [Ruby](https://www.ruby-lang.org/en/)'s global variables and constants in response to cheat sheet searches like *ruby globals*, *ruby global constants*, *ruby global variables*.

**What is the data source for your Instant Answer? (Provide a link if possible)**
The list items are copied verbatim from the [Ruby source][] at [doc/globals.rdoc][].  This document is rendered at http://docs.ruby-lang.org/en/trunk/globals_rdoc.html, which I've used as the answer `sourceUrl` as it is more readable than the raw rdoc markup.

The source lists variables and constants.  I added sections to break the list of variables up; without the addition of a **Long Variables** section there is a lot of whitespace to the right of the one- and two-character variables.

[Ruby source]: https://github.com/ruby/ruby
[doc/globals.rdoc]: https://github.com/ruby/ruby/blob/trunk/doc/globals.rdoc

**Why did you choose this data source?**
This is a document from the Ruby project itself, so should be close to the truth.  I chose http://docs.ruby-lang.org as a reference as it is under the official Ruby domain.

**Are there any other alternative (better) data sources?**
The same data is presented at http://ruby-doc.org/core/doc/globals_rdoc.html, and other Ruby documentation servers.  I wasn't able to find a source that uses SSL.

Ryan Davis lists pre-defined variables as part of [Ruby QuickRef][].

[Ruby QuickRef]: http://www.zenspider.com/Languages/Ruby/QuickRef.html#pre-defined-variables

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Programmers

**Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?**
No

**Are you having any problems? Do you need our help with anything?**
I ran `duckpan server` locally, it took me a while to work out that I needed to restart the server to pick up the new JSON file and aliases, though other changes are loaded while the server is running.  I know now to do this, but it would be helpful if the server was restarted automatically when needed, or printed a warning suggesting so.

**Where did you hear about DuckDuckHack? (For first time contributors)**
I've used the search engine for years, and have gradually become aware of the community around it.  I started hearing about Quack & Hack about a month ago, which I hope to get involved in.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
screen shot 2015-12-13 at 12 02 24

---
https://duck.co/ia/view/ruby_globals_cheat_sheet

### CODE REVIEW

**Comment by daxtheduck** [comment]
@dthjones Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

While the team and the community review your submission, we'd like to make sure you look good! If you'd like a free DuckDuckGo tshirt, please send over the following info to merch@duckduckgo.com, and be sure to put "Dev Merch" in the subject line:

Name:
Street Address:
City:
State / Province:
Zip Code:
Country:
Phone (Non-US Orders only): 
T-shirt size:
GitHub ID:
Link to this Pull Request:

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

File: share/goodie/cheat_sheets/json/ruby-globals.json [hunk 1]
```diff
@@ -0,0 +1,249 @@
+{
+  "id": "ruby_globals_cheat_sheet",
+  "name": "Ruby Globals",
+  "description": "Pre-defined global variables and constants in the Ruby programming language",
+  "metadata": {
+    "sourceName": "Documentation for Ruby",
+    "sourceUrl": "http://docs.ruby-lang.org/en/trunk/globals_rdoc.html"
```
Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: Change source name to "Ruby-Lang"

File: share/goodie/cheat_sheets/json/ruby-globals.json [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        "val": "The typical true value."
+      },
+      {
+        "key": "FALSE",
+        "val": "The false itself."
+      },
+      {
+        "key": "NIL",
+        "val": "The nil itself."
+      },
```
Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @dthjones The val fields of first three global constants can be modified to something like "True value","False value"
and "No value"
@moollaza  what do you say?

Comment (GuiltyDolphin) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @jophab `nil` isn't a 'No value' in Ruby, as [this](http://www.skorks.com/2009/09/true-false-and-nil-objects-in-ruby/) page goes into. I agree though, the wording could be changed a bit, maybe `nil literal` or suchlike?

@dthjones Also, I think lowercase for `nil`, `true` and `false` is preferred in Ruby (in general).

Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @GuiltyDolphin  :+1:  You are right.
Lets change the wording a bit.

**Comment by jophab** [comment]
@dthjones @GuiltyDolphin  I think  many of the val fields dont require a "The" in front.
Some fields have redundant  use of "the".
What about avoiding them?

File: share/goodie/cheat_sheets/json/ruby-globals.json (hunk 2 above)
Comment (dthjones) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: From what I understand (I'm not too familiar with Ruby core), `NIL`, `TRUE` and `FALSE` are the singleton instances of `NilClass`, `TrueClass` and `FalseClass`, similar to doing
```ruby
class NilClass
  # ...
end
NIL = NilClass.new
```
`nil`, `true` and `false` are, I think, "magic" on the syntax level, and equivalent to using the constants directly:
```ruby
nil == NIL     #=> true
true == TRUE   #=> true
false == FALSE #=> true
```
@jophab, @GuiltyDolphin, is the following more comprehensible?

`TRUE`: Singleton instance of TrueClass (the true value)
`FALSE`: Singleton instance of FalseClass (the false value)
`NIL`: Singleton instance of NilClass (the nil value)

**Comment by dthjones** [comment]
@jophab I can re-write this to be more concise and consistent generally. Is it acceptable to deviate from the source?

File: share/goodie/cheat_sheets/json/ruby-globals.json (hunk 2 above)
Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: image

Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @dthjones your new defenition for true false and nil is better than the old one
Lets change to the new one. :+1:

Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @dthjones And consider @GuiltyDolphin  said about lowercase true ,false and nil.

**Comment by jophab** [comment]
@dthjones You can deviate from the source only if you are sure that it does not gives any wrong information.
Its great if the cheat sheet itself can give quality information than source in a better language.

File: share/goodie/cheat_sheets/json/ruby-globals.json (hunk 1 above)
Comment (dthjones) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: Most pages under http://docs.ruby-lang.org/en/ have the title prefix *Documentation for Ruby*, https://www.ruby-lang.org/en/ has the title *Ruby Programming Language*.  I could use that,  "Documentation for Ruby Programming Language" or "docs.ruby-lang.org" for `sourceName` if any of those read better?

Comment (jophab) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @dthjones Source name is used to reflect the domain of the source url.
For example https://en.wikipedia.org/wiki/Ruby_(programming_language) .
here the source name is just  "Wikipedia" . that's  enough. no need of .org or anything.
To clarify your doubts just examine code of some other cheat sheets. :+1:

Comment (dthjones) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @jophab No problem, I've pushed another commit to change this.

File: share/goodie/cheat_sheets/json/ruby-globals.json [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    "Short Variables",
+    "Long Variables",
+    "Command Line Switches",
+    "Global Constants"
+  ],
+  "sections": {
+    "Short Variables": [
+      {
+        "key": "$!",
+        "val": "Last raised exception."
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: @dthjones let's remove the periods `.` -- we don't use them in other cheat sheets typically

**Comment by moollaza** [comment]
@dthjones thanks for contributing and welcome to DuckDuckHack!

> I ran duckpan server locally, it took me a while to work out that I needed to restart the server to pick up the new JSON file and aliases, though other changes are loaded while the server is running. I know now to do this, but it would be helpful if the server was restarted automatically when needed, or printed a warning suggesting so.

Thanks for pointing this out! DuckPAN _should_ be reloading automatically, but I don't recall if we updated it to look for the Cheat Sheet JSON files. If not I'll create a task, should be fairly each to watch that directory.

**Comment by moollaza** [comment]
>I can re-write this to be more concise and consistent generally. Is it acceptable to deviate from the source?

@dthjones that's perfectly fine. As long as we know the content is coming from a trusted reliable source, you're welcome to paraphrase if it helps improve the cheat sheet. The source link is mostly there to provide a place for the developer to go if they need further information.

File: share/goodie/cheat_sheets/json/ruby-globals.json (hunk 3 above)
Comment (dthjones) [inline share/goodie/cheat_sheets/json/ruby-globals.json]: Sure thing, I've removed the periods.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=Ruby Globals Cheat Sheet)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/ruby_globals_cheat_sheet_mobile.png)

**Comment by dthjones** [comment]
@moollaza Thank you!  Contributing to DuckDuckHack has been a great experience so far :heart:

File changes not being tracked matches what I saw, I think I got caught out by adding the file, starting the server and renaming it.  Now I know about it, it isn't a problem.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=Ruby Globals Cheat Sheet)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/ruby_globals_cheat_sheet_mobile.png)

**Comment by moollaza** [comment]
@kotoshenya apologies for the delay! this LGTM :+1:

Thanks a lot for contributing (we really appreciate it :smiley:). We invite you to stick around and comment/help on other PR's & Issues or even submit an Instant Answer!

**Comment by moollaza** [comment]
If you need ideas about how to contribute next, please feel free to reach out to me on Slack! I'm happy to chat about how you can help :+1:

**Comment by moollaza** [comment]
>@moollaza Thank you! Contributing to DuckDuckHack has been a great experience so far :heart:

BTW We're always open to any thoughts or feedback you have on the process, documentation, tools, etc! So if you have anything you'd like us to know, please do tell me!

**Comment by alohaas** [comment]
@kotoshenya  thanks again for creating the Ruby Globals Cheat Sheet! You’ve done a great job here and we'd like you to help our community tackle the next big challenge – our mission to become the best search engine for programmers: https://duck.co/blog/post/297/help-for-programmers.

When you're ready to move on, please: 

- Create a new Instant Answer to increase our Programming IA coverage

OR

- Test live programming Instant Answers (or those in development) for bugs and suggest improvements 

Detailed instructions explaining how you can help are available at the top of this [page](https://github.com/duckduckgo/duckduckgo/wiki/Programming-IA-Coverage).

Thanks again!

**Pull request merged**